sdks/java/io/redis/src/main/java/org/apache/beam/sdk/io/redis/RedisIO.java

```diff
@@ -0,0 +1,294 @@
+     * @return The corresponding {@link Read} {@link PTransform}.
+     */
+    public Read withRedisService(RedisService redisService) {
+      checkArgument(redisService != null, "RedisIO.read().withRedisService(service) called with"
+          + " null service");
+      return builder().setRedisService(redisService).build();
+    }
+
+    @Override
+    public void validate(PBegin input) {
+      checkState(connection() != null || redisService() != null, "RedisIO.read() requires a "
+          + "connection to be set withConnection(connection) or a service to be set "
+          + "withRedisService(service)");
+      checkState(command() != null,  "RedisIO.read() requires a command to be set via "
+          + "withCommand(command)");
+      checkState(keyPattern() != null, "RedisIO.read() requires a key pattern to be set via "
+          + "withKeyPattern(keyPattern");
+    }
+
+    @Override
+    public void populateDisplayData(DisplayData.Builder builder) {
+      if (connection() != null) {
+        connection().populateDisplayData(builder);
+      }
+      builder.addIfNotNull(DisplayData.item("command", command().toString()));
+    }
+
+    @Override
+    public PCollection<KV<String, String>> expand(PBegin input) {
+      return input.getPipeline()
+          .apply(org.apache.beam.sdk.io.Read.from(
+              new RedisSource(
+                  keyPattern(),
+                  new SerializableFunction<PipelineOptions, RedisService>() {
+                    @Override
+                    public RedisService apply(PipelineOptions pipelineOptions) {
+                      return getRedisService(pipelineOptions);
+                    }
+                  })));
+    }
+
+    /**
+     * Helper function to either get a fake/mock Redis service provided by
+     * {@link #withRedisService(RedisService)} or creates and returns an implementation of a
+     * concrete Redis service dealing with an actual Redis server.
+     */
+    @VisibleForTesting
+    RedisService getRedisService(PipelineOptions pipelineOptions) {
+      if (redisService() != null) {
+        return redisService();
+      }
+      return new RedisServiceImpl(connection());
+    }
+
+  }
+
+  /**
+   * A bounded source reading key-value pairs from a Redis server.
+   */
+  // visible for testing
+  protected static class RedisSource extends BoundedSource<KV<String, String>> {
+
+    private final String keyPattern;
+    private final SerializableFunction<PipelineOptions, RedisService> serviceFactory;
+
+    protected RedisSource(String keyPattern,
+                          SerializableFunction<PipelineOptions, RedisService> serviceFactory) {
+      this.keyPattern = keyPattern;
+      this.serviceFactory = serviceFactory;
+    }
+
+    @Override
+    public List<RedisSource> splitIntoBundles(long desiredBundleSizeBytes,
+                                              PipelineOptions pipelineOptions) {
+      // TODO cluster with one source per slot
```

Comment:
In the case of a cluster connection, it would be possible to split the read with one source per node in the cluster.